srml/im-online/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+	SubmitTransaction,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as ImOnline {
+		// The block number when we should gossip.
+		GossipAt get(gossip_at) config(): T::BlockNumber;
+
+		// The session index when the last new era started.
+		LastNewEraStart get(last_new_era_start) config(): Option<session::SessionIndex>;
+
+		// Mapping of AuthorityId to the serialized struct `offchain::LocalNetworkState`.
+		// We currently don't provide the deserialized struct, since the contained types
+		// don't compile for wasm (and since there is currently no need).
+		//
+		// The `StorageMap` implementation currently doesn't provide an easy way to
+		// clear the map, hence we use this workaround of a `double_map` with the
+		// fixed `PREFIX` constant.
+		ReceivedHeartbeats get(received_heartbeats) config(): double_map u32, blake2_256(T::AuthorityId) => Vec<u8>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Number of sessions per era.
+		const SessionsPerEra: session::SessionIndex = T::SessionsPerEra::get();
+
+		fn deposit_event<T>() = default;
+
+		fn heartbeat(
+			origin,
+			heartbeat: Heartbeat<T::BlockNumber, T::AuthorityId>,
+			_signature: Vec<u8>
+		) {
+			ensure_none(origin)?;
+
+			let exists = <ReceivedHeartbeats<T>>::exists(PREFIX, &heartbeat.authority_id);
+			if !exists {
+				let now = <system::Module<T>>::block_number();
+				Self::deposit_event(RawEvent::HeartbeatReceived(now, heartbeat.authority_id.clone()));
+
+				<ReceivedHeartbeats<T>>::insert(PREFIX, &heartbeat.authority_id, heartbeat.network_state);
+			}
+		}
+
+		// Runs after every block.
+		fn offchain_worker(block_number: T::BlockNumber) {
+			fn gossip<T: Trait>(block_number: T::BlockNumber) -> Result<(), OffchainErr> {
+				match sr_io::local_authority_pubkey(<T as Trait>::CRYPTO_KIND) {
+					Ok(key) => {
+						let authority_id = <T as Trait>::AuthorityId::decode(&mut &key[..])
+							.ok_or(OffchainErr::DecodeAuthorityId)?;
+						let network_state =
+							sr_io::local_network_state().map_err(|_| OffchainErr::LocalNetworkState)?;
+						let heartbeat_data = Heartbeat {
+							block_number,
+							network_state,
+							session_index: <session::Module<T>>::current_index(),
+							authority_id,
+						};
+
+						match sr_io::sign(None, <T as Trait>::CRYPTO_KIND, &heartbeat_data.encode()) {
+							Ok(signature) => {
+								let call = Call::heartbeat(heartbeat_data, signature);
+								let ex = T::UncheckedExtrinsic::new_unsigned(call.into())
+									.ok_or(OffchainErr::ExtrinsicCreation)?;
+								sr_io::submit_transaction(&ex)
+									.map_err(|_| OffchainErr::SubmitTransaction)?;
+								Ok(())
+							},
+							Err(_) => Err(OffchainErr::FailedSigning),
+						}
+					},
+					Err(_) => {
+						// we only run when a local authority key is configured
+						Ok(())
+					}
+				}
+			}
+
+			let gossip_at = <GossipAt<T>>::get();
+			let now = <system::Module<T>>::block_number();
```

Comment:
We already have `block_number`, I don't think we need to query `system`. Alternative we can get rid of this parameter for offchain workers.